### PR TITLE
documentation for ensemble parallelism

### PR DESCRIPTION
**Pull request by colinjcotter**

### CODE REVIEW

File: docs/source/parallelism.rst [hunk 1]
```diff
@@ -75,5 +75,62 @@ different simulations on the two halves we would write.
...
+
+.. code-block:: python
+
+   from firedrake import *
+
+   my_ensemble = Ensemble(COMM_WORLD, 4)
+
+Then, the spatial sub-communicator must be passed to ``Mesh`` (or via
+inbuilt mesh types), so that it will then be used by function spaces
+and functions derived from the mesh.
```
Comment (dham) [inline docs/source/parallelism.rst]: You're missing the code for the construction of the mesh, which appears to be important at this stage.

File: docs/source/parallelism.rst [hunk 2]
```diff
@@ -75,5 +75,62 @@ different simulations on the two halves we would write.
...
+.. code-block:: python
+
+    q = Constant(manager.ensemble_comm.rank + 1)
+    u.interpolate(sin(q*pi*x)*cos(q*pi*y))
+
+MPI communications across the spatial sub-communicator (i.e., within
+an ensemble member) are handled automatically by Firedrake, whilst MPI
+communications across the ensemble sub-communicator (i.e., between ensemble
+members) are handled through methods of ``Ensemble``. Currently only
+global reductions are supported.
```
Comment (dham) [inline docs/source/parallelism.rst]: I think this is cryptic. I don't think that anyone who doesn't already understand how our ensemble communicators work will understand it after this either.

**Comment by colinjcotter** [comment]
OK, I have updated this PR with some hopefully better description
and a figure, and correcting the missing mesh and bad naming.

File: docs/source/parallelism.rst [hunk 3]
```diff
@@ -75,5 +75,71 @@ different simulations on the two halves we would write.
...
+
+.. code-block:: python
+
+    q = Constant(my_ensemble.ensemble_comm.rank + 1)
+    u.interpolate(sin(q*pi*x)*cos(q*pi*y))
+
+MPI communications across the spatial sub-communicator (i.e., within
+an ensemble member) are handled automatically by Firedrake, whilst MPI
+communications across the ensemble sub-communicator (i.e., between ensemble
+members) are handled through methods of ``Ensemble``. Currently only
```
Comment (dham) [inline docs/source/parallelism.rst]: See http://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#python-roles for cross-referencing syntax. In this case you want:

    :class:`~.Ensemble`

**Comment by colinjcotter** [comment]
Can we merge now?

**Comment by dham** [comment]
Yes

Dr David Ham
Department of Mathematics
Imperial College London

On 4 Sep 2018, at 13:42, Colin J Cotter <notifications@github.com<mailto:notifications@github.com>> wrote:


Can we merge now?

**Pull request merged**